In [1]:
# start pandas
import pandas as pd

In [2]:
# access CSV files
school_csv = "Resources/schools_complete.csv"
student_csv = "Resources/students_complete.csv"
#load CSV to pandas DataFrames
school_df = pd.read_csv(school_csv)
student_df = pd.read_csv(student_csv)


In [3]:
#Merge school and student DataFrames
unified_df=pd.merge(student_df, school_df, how="left", on=["school_name","school_name"])


# District Overall Summary

In [4]:
#Create blank dictionary for district summary
dist_dict={}

In [5]:
#Calculate Total number of Schools and add to dictionary
sch_count=len(unified_df["school_name"].unique())
dist_dict["Total Schools"]=sch_count


In [6]:
#Calculate Total number of Students and add to dictionary
stu_count=len(student_df["Student ID"].unique())
dist_dict["Total Students"]=stu_count



In [7]:
#Calculate Total Budget and add to dictionary
total_budget=school_df.budget.sum()
dist_dict["Total Budget"]=total_budget


In [8]:
#Calculate Average Math Score and add to dictionary
avg_math=student_df.math_score.mean()
dist_dict["Average Math Score"]=avg_math


In [9]:
#Calculate Average Reading Score and add to Dictionary
avg_read=student_df.reading_score.mean()
dist_dict["Average Reading Score"]=avg_read


In [10]:
#Calculate percentage of students who passed Math and add to dictionary
math_pass=unified_df[(unified_df["math_score"] >= 70)].count()["student_name"]
math_perc = math_pass / float(stu_count) * 100
dist_dict["% Passing Math"]=math_perc


In [11]:
#Calculate percentage of students who passed reading and add to dictionary
read_pass=unified_df[(unified_df["reading_score"] >= 70)].count()["student_name"]
read_perc=read_pass/float(stu_count)*100
dist_dict["% Passing Reading"]=read_perc


In [12]:
#Calculate percentage of students who passed both and add to dictionary
both_pass = unified_df[(unified_df["math_score"] >= 70) & (unified_df["reading_score"] >= 70)].count()["student_name"]
both_perc = both_pass /  float(stu_count) * 100
dist_dict["% OverallPassing"]=both_perc


In [13]:
#Create raw DataFrame of Overall District Metrics
district_df=pd.DataFrame(dist_dict, index=[0])


In [14]:
#Format currency and add thousands separation for final presentation
district_summary=district_df.copy()
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)
district_summary

Total Schools Total Students    Total Budget  Average Math Score  \
0             15         39,170  $24,649,428.00           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % OverallPassing  
0               81.87784       74.980853          85.805463         65.172326

# Per School Summary 

In [15]:
#Set school name as the index
type_df=school_df.set_index(["school_name"])["type"]


In [16]:
#determine number of students per school
stu_per_sch=unified_df["school_name"].value_counts()


In [17]:
#determine total budget per school
bud_per_sch=unified_df.groupby(["school_name"]).mean()["budget"]


In [18]:
#determine budget per capita per school
bud_per_cap=bud_per_sch/stu_per_sch


In [19]:
#determine average math score per school
per_sch_math=unified_df.groupby(["school_name"]).mean()["math_score"]


In [20]:
#determine average reading score per school
per_sch_read=unified_df.groupby(["school_name"]).mean()["reading_score"]


In [21]:
#create filtered DataFrame of students passing math
stu_pass_math_df=unified_df[(unified_df["math_score"]>=70)]


In [22]:
#create filtered DataFrame of students passing reading
stu_pass_read_df=unified_df[(unified_df["reading_score"]>=70)]


In [23]:
#create filtered DataFrame of students who passed both math and reading
stu_pass_both_df = unified_df[(unified_df["reading_score"] >= 70) & (unified_df["math_score"] >= 70)]


In [24]:
#determine passing rate for math per school
per_sch_math_perc = stu_pass_math_df.groupby(["school_name"]).count()["student_name"] / stu_per_sch * 100


In [25]:
per_sch_read_perc = stu_pass_read_df.groupby(["school_name"]).count()["student_name"] / stu_per_sch * 100


In [26]:
per_sch_both_perc = stu_pass_both_df.groupby(["school_name"]).count()["student_name"] / stu_per_sch * 100


In [27]:
#Combine information into raw Per School Summary DataFrame
per_sch_sum=pd.DataFrame({"School Type":type_df,
                        "Total Students":stu_per_sch,
                        "Total School Budget":bud_per_sch,
                        "Per Student Budget":bud_per_cap,
                        "Average Math Score":per_sch_math,
                        "Average Reading Score":per_sch_read,
                        "% Passing Math":per_sch_math_perc,
                        "% Passing Reading":per_sch_read_perc,
                        "% Overall Passing":per_sch_both_perc})


In [28]:
#Add Formatting for thousands separation and currency
per_school_summary=per_sch_sum.copy()
per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)
per_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Bailey High School               $628.00           77.048432   
Cabrera High School              $582.00           83.061895   
Figueroa High School             $639.00           76.711767   
Ford High School                 $644.00           77.102592   
Griffin High School              $625.00           83.351499   
Hernandez High School            $652.00           77.289752   
Holden High School               $581.00           83.803279   
Huang High School                $655.00           76.629414   
Johnson High School              $650.00           77.072464   
Pena High School                 $609.00           83.839917   
Rodriguez High School            $637.00           76.842711   
Shelton High School              $600.00           83.359455   
Thomas High School               $638.00           83.418349   
Wilson High School               $578.00           83.274201   
Wright High School               $583.00           83.682222   

                       Average Reading Score  % Passing Math  \
Bailey High School                 81.033963       66.680064   
Cabrera High School                83.975780       94.133477   
Figueroa High School               81.158020       65.988471   
Ford High School                   80.746258       68.309602   
Griffin High School                83.816757       93.392371   
Hernandez High School              80.934412       66.752967   
Holden High School                 83.814988       92.505855   
Huang High School                  81.182722       65.683922   
Johnson High School                80.966394       66.057551   
Pena High School                   84.044699       94.594595   
Rodriguez High School              80.744686       66.366592   
Shelton High School                83.725724       93.867121   
Thomas High School                 83.848930       93.272171   
Wilson High School                 83.989488       93.867718   
Wright High School                 83.955000       93.333333   

                       % Passing Reading  % Overall Passing  
Bailey High School             81.933280          54.642283  
Cabrera High School            97.039828          91.334769  
Figueroa High School           80.739234          53.204476  
Ford High School               79.299014          54.289887  
Griffin High School            97.138965          90.599455  
Hernandez High School          80.862999          53.527508  
Holden High School             96.252927          89.227166  
Huang High School              81.316421          53.513884  
Johnson High School            81.222432          53.539172  
Pena High School               95.945946          90.540541  
Rodriguez High School          80.220055          52.988247  
Shelton High School            95.854628          89.892107

# Schools by Performance

In [29]:
#Top 5 Schools by % Overall Passing
top_schools=per_school_summary.sort_values("% Overall Passing", ascending=False)
top_schools.head(5)

School Type  Total Students Total School Budget  \
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget  Average Math Score  \
Cabrera High School            $582.00           83.061895   
Thomas High School             $638.00           83.418349   
Griffin High School            $625.00           83.351499   
Wilson High School             $578.00           83.274201   
Pena High School               $609.00           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

In [30]:
#Bottom 5 Schools by % Overall Passing
bottom_schools=per_school_summary.sort_values("% Overall Passing")
bottom_schools.head(5)

School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget  Average Math Score  \
Rodriguez High School            $637.00           76.842711   
Figueroa High School             $639.00           76.711767   
Huang High School                $655.00           76.629414   
Hernandez High School            $652.00           77.289752   
Johnson High School              $650.00           77.072464   

                       Average Reading Score  % Passing Math  \
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

# Scores by Grade

In [31]:
#Separate Data by Grade
ninth_df = unified_df[(unified_df["grade"] == "9th")]
tenth_df = unified_df[(unified_df["grade"] == "10th")]
eleventh_df = unified_df[(unified_df["grade"] == "11th")]
twelfth_df = unified_df[(unified_df["grade"] == "12th")]


In [32]:
#group by average scores
ninth_scores=ninth_df.groupby(["school_name"]).mean()
tenth_scores=tenth_df.groupby(["school_name"]).mean()
eleventh_scores=eleventh_df.groupby(["school_name"]).mean()
twelfth_scores=twelfth_df.groupby(["school_name"]).mean()

In [33]:
#select math and reading scores
#math
ninth_math=ninth_scores["math_score"]
tenth_math=tenth_scores["math_score"]
eleventh_math=eleventh_scores["math_score"]
twelfth_math=twelfth_scores["math_score"]
#reading
ninth_read=ninth_scores["reading_score"]
tenth_read=tenth_scores["reading_score"]
eleventh_read=eleventh_scores["reading_score"]
twelfth_read=twelfth_scores["reading_score"]

In [34]:
#math scores by grade DataFrame
math_scores_by_grade=pd.DataFrame({"9th":ninth_math,
                               "10th":tenth_math,
                               "11th":eleventh_math,
                               "12th":twelfth_math})
#remove "school_name" formatting
math_scores_by_grade.index.name=None

math_scores_by_grade

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

In [35]:
#reading scores by grade DataFrame
reading_scores_by_grade=pd.DataFrame({"9th":ninth_read,
                               "10th":tenth_read,
                               "11th":eleventh_read,
                               "12th":twelfth_read})
#remove "school_name" formatting
reading_scores_by_grade.index.name=None


reading_scores_by_grade

9th       10th       11th       12th
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

# Scores by Spending

In [36]:
#define bins to sort per capita spending
per_cap_bins = [0, 585, 630, 645, 680]
spend_labels = ["<$585", "$585-630", "$630-645", "$645-680"]

In [37]:
#copy un-formatted "per school" summary for sorting
spend_df=per_sch_sum.copy()


In [38]:
#Add categories for spending ranges
spend_df["Spending Ranges (Per Student)"]=pd.cut(spend_df["Per Student Budget"],per_cap_bins,labels=spend_labels, include_lowest=True)


In [39]:
#Determine performance averages for the spending ranges 
spend_math = spend_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Math Score"]
spend_read = spend_df.groupby(["Spending Ranges (Per Student)"]).mean()["Average Reading Score"]
spend_pass_math = spend_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Math"]
spend_pass_read = spend_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Passing Reading"]
spend_pass_both = spend_df.groupby(["Spending Ranges (Per Student)"]).mean()["% Overall Passing"]

In [40]:
#Assemble Spending Range DataFrame
spending_summary=pd.DataFrame({"Average Math Score":spend_math,
                            "Average Reading Score":spend_read,
                            "% Passing Math":spend_pass_math,
                            "% Passing Reading":spend_pass_read,
                            "% Overall Passing":spend_pass_both})
spending_summary

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                   83.455399              83.933814   
$585-630                                81.899826              83.155286   
$630-645                                78.518855              81.624473   
$645-680                                76.997210              81.027843   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                               93.460096          96.610877   
$585-630                            87.133538          92.718205   
$630-645                            73.484209          84.391793   
$645-680                            66.164813          81.133951   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                  90.369459  
$585-630                               81.418596  
$630-645                               62.857656  
$645-680                               53.526855

# Scores by Size

In [41]:
#define bins to sort by school size
size_bins = [0, 1000, 2000, 5000]
size_labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [42]:
#copy un-formatted "per school" summary for sorting
size_df=per_sch_sum.copy()

In [43]:
#Add categories for size ranges
size_df["School Size"]=pd.cut(size_df["Total Students"],size_bins,labels=size_labels, include_lowest=True)


In [44]:
#Determine performance averages for school sizes
size_math = size_df.groupby(["School Size"]).mean()["Average Math Score"]
size_read = size_df.groupby(["School Size"]).mean()["Average Reading Score"]
size_pass_math = size_df.groupby(["School Size"]).mean()["% Passing Math"]
size_pass_read = size_df.groupby(["School Size"]).mean()["% Passing Reading"]
size_pass_both = size_df.groupby(["School Size"]).mean()["% Overall Passing"]

In [45]:
#Assemble Size Range DataFrame
size_summary=pd.DataFrame({"Average Math Score":size_math,
                            "Average Reading Score":size_read,
                            "% Passing Math":size_pass_math,
                            "% Passing Reading":size_pass_read,
                            "% Overall Passing":size_pass_both})
size_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821598              83.929843       93.550225   
Medium (1000-2000)           83.374684              83.864438       93.599695   
Large (2000-5000)            77.746417              81.344493       69.963361   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)               96.099437          89.883853  
Medium (1000-2000)          96.790680          90.621535  
Large (2000-5000)           82.766634          58.286003


# Scores by Type

In [46]:
#Determine performance averages for school types
type_math = per_sch_sum.groupby(["School Type"]).mean()["Average Math Score"]
type_read = per_sch_sum.groupby(["School Type"]).mean()["Average Reading Score"]
type_pass_math = per_sch_sum.groupby(["School Type"]).mean()["% Passing Math"]
type_pass_read = per_sch_sum.groupby(["School Type"]).mean()["% Passing Reading"]
type_pass_both = per_sch_sum.groupby(["School Type"]).mean()["% Overall Passing"]
type_size=per_sch_sum.groupby(["School Type"]).mean()["Total Students"]
type_size

School Type
Charter     1524.250000
District    3853.714286
Name: Total Students, dtype: float64

In [47]:
#create Data Frame to compare student performance between school types
type_summary=pd.DataFrame({"Average Math Score":type_math,
                            "Average Reading Score":type_read,
                            "% Passing Math":type_pass_math,
                            "% Passing Reading":type_pass_read,
                            "% Overall Passing":type_pass_both})
type_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473852              83.896421       93.620830   
District              76.956733              80.966636       66.548453   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586489          90.432244  
District             80.799062          53.672208

# Summary of Analysis

The district has an average math score in the high 70s (78.99%) and a reading score in the low 80s (81.88%). Though Reading seems to be in better shape with 85.81% of students passing compared to 74.98% passing math, the district has a dismal 65.17% of students successfully passing both. 
There seem to be no major variances in the various grades per school. Student Performance also does not improve with higher per capita spending; in fact higher average spending generally correlates with lower percentages of students passing both subjects. There is also a strong correlation with large schools (>2000 students) performing dramatically worse (58.29% passing both) than to medium and small schools (90.62% and 89.88% passing both subjects, respectively).
Charter schools also out performed district schools by a significant margin. The top 5 schools were all charter schools, and the bottom 5 were all district schools. 90.43% of charter students passed both subjects compared to only 53.67% of district students. It is worth noting, however, that charter schools also had a lower average number of students (1524.25) than district schools (3853.71). It is fair to ask if there is something about the charter style of education that improves results or if it is simply the reduced number of students. 